dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerVectorStoreRecordCollection.cs

```diff
@@ -429,10 +428,7 @@ public async IAsyncEnumerable<TKey> UpsertAsync(IEnumerable<TRecord> records,
 
         var keyProperty = this._model.KeyProperty;
 
-        foreach (var record in records)
-        {
-            yield return (TKey)keyProperty.GetValueAsObject(record!)!;
-        }
+        return records.Select(r => (TKey)keyProperty.GetValueAsObject(r!)!).ToList();
```

Comment:
is the bang still required?

Reply:
The one on `r` isn't (because we now have a `notnull` constraint on TRecord), the other unfortunately still is... We need to replace at least some of the calls to GetValueAsObject (and Set) with generic getters/setters - that would allow #11183, but would also obviate the bang (since TKey is constrained to `notnull`. Later :)